cpp/src/arrow/compute/kernels/filter.cc

```diff
@@ -15,44 +15,468 @@
+                                       dict_array->dictionary(), out);
+  }
+
+ private:
+  std::unique_ptr<FilterKernel> impl_;
+};
+
+template <>
+class FilterImpl<ExtensionType> : public FilterKernel {
+ public:
+  FilterImpl(const std::shared_ptr<DataType>& type, std::unique_ptr<FilterKernel> impl)
+      : FilterKernel(type), impl_(std::move(impl)) {}
+
+  Status Filter(FunctionContext* ctx, const Array& values, const BooleanArray& filter,
+                int64_t length, std::shared_ptr<Array>* out) override {
+    auto ext_array = checked_cast<const ExtensionArray*>(&values);
+    // To take from an extension array, apply the current kernel to storage.
+    std::shared_ptr<Array> taken_storage;
+    RETURN_NOT_OK(
+        impl_->Filter(ctx, *ext_array->storage(), filter, length, &taken_storage));
+    *out = ext_array->extension_type()->MakeArray(taken_storage->data());
+    return Status::OK();
+  }
+
+ private:
+  std::unique_ptr<FilterKernel> impl_;
+};
+
+Status FilterKernel::Make(const std::shared_ptr<DataType>& value_type,
+                          std::unique_ptr<FilterKernel>* out) {
+  switch (value_type->id()) {
+#define NO_CHILD_CASE(T)                                           \
+  case T##Type::type_id:                                           \
+    *out = internal::make_unique<FilterImpl<T##Type>>(value_type); \
+    return Status::OK()
+
+#define SINGLE_CHILD_CASE(T, CHILD_TYPE)                                                \
+  case T##Type::type_id: {                                                              \
+    auto t = checked_pointer_cast<T##Type>(value_type);                                 \
+    std::unique_ptr<FilterKernel> child_filter_impl;                                    \
+    RETURN_NOT_OK(FilterKernel::Make(t->CHILD_TYPE(), &child_filter_impl));             \
+    *out = internal::make_unique<FilterImpl<T##Type>>(t, std::move(child_filter_impl)); \
+    return Status::OK();                                                                \
+  }
+
+    NO_CHILD_CASE(Null);
+    NO_CHILD_CASE(Boolean);
+    NO_CHILD_CASE(Int8);
+    NO_CHILD_CASE(Int16);
+    NO_CHILD_CASE(Int32);
+    NO_CHILD_CASE(Int64);
+    NO_CHILD_CASE(UInt8);
+    NO_CHILD_CASE(UInt16);
+    NO_CHILD_CASE(UInt32);
+    NO_CHILD_CASE(UInt64);
+    NO_CHILD_CASE(Date32);
+    NO_CHILD_CASE(Date64);
+    NO_CHILD_CASE(Time32);
+    NO_CHILD_CASE(Time64);
+    NO_CHILD_CASE(Timestamp);
+    NO_CHILD_CASE(Duration);
+    NO_CHILD_CASE(HalfFloat);
+    NO_CHILD_CASE(Float);
+    NO_CHILD_CASE(Double);
+    NO_CHILD_CASE(String);
+    NO_CHILD_CASE(Binary);
+    NO_CHILD_CASE(FixedSizeBinary);
+    NO_CHILD_CASE(Decimal128);
+
+    SINGLE_CHILD_CASE(Dictionary, index_type);
+    SINGLE_CHILD_CASE(Extension, storage_type);
+
+    NO_CHILD_CASE(List);
+    NO_CHILD_CASE(FixedSizeList);
+    NO_CHILD_CASE(Map);
+
+    case Type::STRUCT: {
+      std::vector<std::unique_ptr<FilterKernel>> child_kernels;
```

Comment:
I just realized you can't do this, you need to account for the top-level parent null bitmap.

Reply:
@fsaintjacques I do; it's built alongside the child fields in `FilterImpl<StructType>`: https://github.com/apache/arrow/pull/4366/files/5a224b24ced09ed6692e0d0ef73a399dab072229#diff-4cce1db37d9827bc657384ea82da0d63R182